wcfsetup/install/files/lib/system/search/SearchHandler.class.php

```diff
@@ -0,0 +1,348 @@
+
+        return $this->saveSearch();
+    }
+
+    private function initParameters(): void
+    {
+        if (empty($this->parameters['sortField'])) {
+            $this->parameters['sortField'] = SEARCH_DEFAULT_SORT_FIELD;
+        }
+        if (empty($this->parameters['sortOrder'])) {
+            $this->parameters['sortOrder'] = SEARCH_DEFAULT_SORT_ORDER;
+        }
+    }
+
+    private function buildConditions(): void
+    {
+        $this->initObjectTypeNames();
+        $this->initConditionBuilder();
+
+        $this->buildUserCondition();
+        $this->buildDateCondition();
+        $this->buildLanguageCondition();
+        $this->buildTypeBasedConditionBuilders();
+    }
+
+    private function initObjectTypeNames(): void
+    {
+        if (!empty($this->parameters['type'])) {
+            $this->objectTypeNames[] = $this->parameters['type'];
+        } else {
+            $this->objectTypeNames = \array_keys(SearchEngine::getInstance()->getAvailableObjectTypes());
+        }
+    }
+
+    private function initConditionBuilder(): void
+    {
+        $conditionBuilderClassName = SearchEngine::getInstance()->getConditionBuilderClassName();
+        $this->conditionBuilder = new $conditionBuilderClassName(false);
+    }
+
+    private function buildUserCondition(): void
+    {
+        $userIDs = $this->getUserIDs();
+        if (!empty($userIDs)) {
+            $this->conditionBuilder->add('userID IN (?)', [$userIDs]);
+        }
+    }
+
+    private function buildDateCondition(): void
+    {
+        $startDate = !empty($this->parameters['startDate']) ? @\strtotime($this->parameters['startDate']) : 0;
+        $endDate = !empty($this->parameters['endDate']) ? @\strtotime($this->parameters['endDate']) : 0;
+        if ($startDate && $endDate) {
+            $this->conditionBuilder->add('time BETWEEN ? AND ?', [$startDate, $endDate]);
+        } elseif ($startDate) {
+            $this->conditionBuilder->add('time > ?', [$startDate]);
+        } elseif ($endDate) {
+            $this->conditionBuilder->add('time < ?', [$endDate]);
+        }
+    }
+
+    private function buildLanguageCondition(): void
+    {
+        if (
+            !empty($this->parameters['q'])
+            && LanguageFactory::getInstance()->multilingualismEnabled()
+            && \count(WCF::getUser()->getLanguageIDs())
+        ) {
+            $this->conditionBuilder->add(
+                '(languageID IN (?) OR languageID = 0)',
+                [WCF::getUser()->getLanguageIDs()]
+            );
+        }
+    }
+
+    private function buildTypeBasedConditionBuilders(): void
+    {
+        $form = $this->getSearchFormEmulation();
+
+        foreach ($this->objectTypeNames as $key => $objectTypeName) {
+            $objectType = SearchEngine::getInstance()->getObjectType($objectTypeName);
+            if ($objectType === null) {
+                throw new SystemException('unknown search object type ' . $objectTypeName);
```

Comment:
No SystemException in new code, please. `$this->objectTypeNames` should probably be validated at the point where untrusted information may be inserted (i.e. initObjectTypeNames).